### PR TITLE
Add Python scripting interface

### PR DESCRIPTION
**Pull request by TsarFox**
Hello,

I've implemented a scripting interface for toxic - a feature present in many IRC and XMPP clients that I felt would be a useful addition to toxic. Similar to video and audio support, it can be disabled at compile time. The API defines primitives suitable for basic automation, such as an interface to toxic's built-in commands and the ability to register new, custom commands. Additionally, this pull request provides a means of specifying scripts to be run when toxic starts from the config file.

The interface is strictly python3 as of now; however the code should be compatible with python2 if you wish to change it. I decided on python3 as python2's EOL is approaching, and decided against giving a compile-time choice between the two to ensure script compatibility.

The Python interface is separated from general API functionality, so future integration with other scripting languages should not be hard to implement.

### CODE REVIEW

**Comment by JFreegman** [comment]
This might take a while for me to review. For starters, could you run the C portions of the code through astyle using the astylerc script in the base directory?

**Comment by TsarFox** [comment]
Sure thing

**Comment by GrayHatter** [comment]
Currently reviewing

@endoffile78 @robinlinden you guys know python too right?

File: src/python_api.c [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        free(python_commands.name);
+
+    for (cur = python_commands.next; cur != NULL;) {
+        old = cur;
+        cur = cur->next;
+        free(old->name);
+        free(old);
+    }
+
+    Py_FinalizeEx();
```
Comment (JFreegman) [inline src/python_api.c]: This function causes compilation to fail for any python version below 3.6. Is there any possible workaround?

File: src/api.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+        line_info_add(self, NULL, NULL, NULL, SYS_MSG, 0, 0, error_str);
+        return;
+    }
+
+    run_python(fp, argv[1]);
+    fclose(fp);
+}
+
+void invoke_autoruns(WINDOW *window, ToxWindow *self)
```
Comment (JFreegman) [inline src/api.c:167]: If this function fails it should alert the user. Both if the directory doesn't exist, and if an invalid path is found.

File: src/python_api.c (hunk 1 above)
Comment (robinlinden) [reply src/python_api.c]: `Py_Finalize()` should work for everything Python 3.

**Comment by TsarFox** [comment]
Both issues should be fixed by the most recent commit

File: src/api.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    status[len] = '\0';
+    return (char *) status;
+}
+
+void api_send(const char *msg)
+{
+    if (msg == NULL || self_window->chatwin->cqueue == NULL)
+        return;
+
+    char *name = api_get_nick();
```
Comment (e0ff) [inline src/api.c:95]: Make sure name isn't NULL

File: src/python_api.c [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    struct python_registered_func *cur;
+    size_t      command_len, help_len;
+    const char *command, *help;
+    PyObject   *callback;
+
+    if (!PyArg_ParseTuple(args, "ssO:register_command", &command, &help, &callback))
+        return NULL;
+
+    if (!PyCallable_Check(callback)) {
+        PyErr_SetString(PyExc_TypeError, "Parameter must be callable");
```
Comment (e0ff) [inline src/python_api.c]: Callback Parameter

File: src/api.c [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    self_window = get_active_window();
+    get_time_str(timefrmt, sizeof(timefrmt));
+
+    line_info_add(self_window, timefrmt, name, NULL, OUT_MSG, 0, 0, "%s", msg);
+    cqueue_add(self_window->chatwin->cqueue, msg, strlen(msg), OUT_MSG,
+               self_window->chatwin->hst->line_end->id + 1);
+    free(name);
+}
+
+void api_execute(const char *input, int mode)
```
Comment (e0ff) [inline src/api.c:112]: It might be better if mode was an enum.

Comment (TsarFox) [reply src/api.c:112]: The enum used in `execute` doesn't have an identifier, so I can't use it as a type without touching the definition. I've added constants to the Python interface for it, though.

**Comment by JFreegman** [comment]
INSTALL.md needs to be updated with the python dependency and new compilation option.

**Comment by TsarFox** [comment]
Updated, and I've got it compiling on a Debian test environment.

**Comment by alexeysvrv** [comment]
I propose to make this option optional. And also turn off the default build. This draws additional dependencies, and scripting is not needed by an ordinary user.

File: src/python_api.c [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+ *
+ */
+
+#include <Python.h>
+
+#include "api.h"
+
+extern Tox       *user_tox;
+
+struct python_registered_func {
```
Comment (robinlinden) [inline src/python_api.c]: Looks like this could be static.

**Review by robinlinden — APPROVED** [review]
I think this all looks good and ready to be merged.

**Comment by JFreegman** [comment]
Any more thoughts on whether this should be compiled by default? I tend to agree with @alexeysvrv that it's a very large dependency for something that presumably a small fraction of users will make use of.

**Comment by TsarFox** [comment]
I think that it's fair to have it disabled by default. Python 3 isn't present on as many systems as say, Perl, in the case of Irssi.

**Comment by JFreegman** [comment]
Ok let's do that then, looks ready to merge otherwise.

**Comment by GrayHatter** [comment]
Reviewed 24 of 28 files at r1, 1 of 3 files at r2, 2 of 3 files at r5, 1 of 3 files at r8.
Review status: 26 of 29 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/jfreegman/toxic/446)*

**Comment by JFreegman** [comment]
I found a couple bugs while testing. 

1. When I run `/fortune 1` the message is received on the opposite end, but toxic is not receiving the acknowledgment from the other client, so a little x shows up, wrongly indicating that the message wasn't sent. If you can't figure this one out I can take a look at it when I have more time.

2. When I run `/fortune <large number>` the toxic process is killed. It also allows negative numbers with an empty message. Should probably limit the input to `> 0 && <= 20` or something.

**Comment by TsarFox** [comment]
I'll take another shot at a fix for the first issue tomorrow when I can set up a better debugging environment.

File: src/api.c [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+void invoke_autoruns(WINDOW *window, ToxWindow *self)
+{
+    struct dirent *dir;
+    char    abspath_buf[PATH_MAX + 1], err_buf[PATH_MAX + 1];
+    size_t  path_len;
+    DIR    *d = opendir(user_settings->autorun_path);
+    FILE   *fp;
+
+    if (d == NULL) {
```
Comment (JFreegman) [inline src/api.c]: This prints an error when the path is never set by the user.

**Review by e0ff — APPROVED** [review]

**Comment by TsarFox** [comment]
@JFreegman If you could look into the issue of messages being wrongly marked as unsent after all, it would be greatly appreciated. api_send, to my understanding, should be modifying the message queue in the same way that chat_onKey does. I'm lost at where else to look.

**Comment by JFreegman** [comment]
I've identified the problem but it's kinda complicated and has to do with the hacky interface, so I'll just merge this and fix it later.

**Comment by JFreegman** [comment]
I've identified the problem but it's kinda complicated and has to do with the hacky interface, so I'll just merge this and fix it later (after you fix the fortune input issues).

If you're curious, it has to do with the fact that api_send is being called in a loop, which doesn't properly update message ID's.

**Pull request merged**